Users can join bets

YarekTenko opened the pull request:
Pivotal Links: 
https://www.pivotaltracker.com/story/show/140512067
https://www.pivotaltracker.com/story/show/141507823

Dependent on #35  
Each BetOption now has a "Bet!" button, which takes a user to the page where they can enter their betting amount and place it or cancel and return to the Bet page.

LeonTenko committed:
Add UserBets controller

-user_bets is a subresource of bets

LeonTenko committed:
Add "Bet!" button for options in Bet

Added the ability to select an option and input the desired betting
amount.

LeonTenko committed:
Define create method for UserBetsController

YarekTenko commented on app/views/bets/show.html.erb:
```diff
@@ -1,49 +1,41 @@
+      </div>
     </div>
   </div>
-  <hr>
```
This section has been deleted in one of the previous PR's already, so it should get resolved upon merging.

qr8r commented on app/views/shared/_bet_display.html.erb:
```diff
@@ -0,0 +1,22 @@
+<% option ||= nil %>
```
Nice!

qr8r commented on app/controllers/user_bets_controller.rb:
```diff
@@ -1,2 +1,9 @@
 class UserBetsController < ApplicationController
+  def new
+    @option = BetOption.find(params[:option_id]) if params[:option_id]
```
`@user_bet.find_or_initialize_by(stuff)`

qr8r commented on app/views/user_bets/new.html.erb:
```diff
@@ -0,0 +1,29 @@
+<% if current_user && @option %>
```
Checking for a logged in user should happen at the controller level.
Could we also make this work if the option doesnt exist

qr8r commented on app/views/user_bets/new.html.erb:
```diff
@@ -0,0 +1,29 @@
+        </div>
+        <%= form_for [@bet, @user_bet] do |f| %>
+          <%= f.hidden_field :user_id, value: current_user.id %>
+          <%= f.hidden_field :bet_option_id, value: @option.id %>
```
Could be a select box, or radio buttons with the specified option selected

qr8r commented on app/views/user_bets/new.html.erb:
```diff
@@ -0,0 +1,29 @@
+          Enter your betting amount
+        </div>
+        <%= form_for [@bet, @user_bet] do |f| %>
+          <%= f.hidden_field :user_id, value: current_user.id %>
```
Using the find_or_initialize above will make it so you dont have to specify the values for all of these fields

qr8r commented on app/controllers/user_bets_controller.rb:
```diff
@@ -5,5 +5,27 @@ def new
 
-  def create; end
+  def create
+    @user_bet = UserBet.new(user_bet_params)
```
Find or initialize again

qr8r commented:
@YarekTenko Its hard to review this pr since the creator relationship is not defined on Bet.

I will pulll it down and review once that gets merged in

LeonTenko committed:
Highlight selected options

The options can now be selected by clicking the row, instead of the
radio button.

qr8r commented:
Jenkins test this please

cbrunsdon commented:
jenkins test this please

jacquesporveau commented:
This branch is out of date with the base branch. Please update.

qr8r commented:
@YarekTenko  still seeing undefined method `creator' for #<Bet:0x007fb185949768> when attempting to test this

ghost commented:
@brendandeere Thanks, Brendan, it should be fixed now.

qr8r commented:
https://www.pivotaltracker.com/story/show/140512067/comments/167080481 

I would like to see this implemented prior to merging ^^

jacquesporveau commented:
This branch is out of date with master. Please update.

LeonTenko committed:
Create BetOptionsController

-BetOptionsController is a subresource of Bet.

LeonTenko committed:
Define new method for BetOptionsController

LeonTenko committed:
Define create method for BetOption

LeonTenko committed:
Add button to create BetOption for Bet

LeonTenko committed:
Stylize headers for Card divs in bet#show

Changed the default grey to black with white text font.

qr8r commented:
placing a user bet with no amount selected gives `nil can't be coerced into BigDecimal`
placing a bet with no option selected gives the cryptic message `Couldn't place your bet`

qr8r commented on app/views/bets/show.html.erb:
```diff
@@ -1,56 +1,77 @@
+              </div>
+            </div>
+            <%= f.submit 'Bet!', class: 'btn btn-success',
+              role: 'button', style: 'width: 100%;', method: 'get' %>
```
Method should go on the form element
:-1: inline styles. Styles should always be written in external style sheets

qr8r commented on config/routes.rb:
```diff
@@ -8,6 +8,8 @@
 
-  resources :bets
+  resources :bets do
+    resources :user_bets, shallow: true
```
You can change this line so that you dont need to include so many empty views

ghost commented:
Wasn't verification a separate story?

qr8r commented:
@YarekTenko Possibly, we have a few stories that specify validation. The big thing for me is if you submit a user be with no amount it breaks the bet/show page. 

If you want to resolve the empty text issue by creating a separate story for that, Im on board. But I wont merge a PR that actually breaks a page

ghost commented:
@brendandeere Ah, gotcha, thanks, Brendan! Hope I fixed it this time.

qr8r commented:
Submitting an empty form leads to a 500 `Render and/or redirect were called multiple times in this action.` 

Please Try and break your own code prior to making a pull request. Test the interactions, enter good values, enter incomplete values, enter invalid values, enter no values, click buttons multiple times, etc

YarekTenko commented:
Woops, accidentally merged #66 into it, sorry, I thought it was based on master.

jacquesporveau commented on app/views/bets/show.html.erb:
```diff
@@ -1,56 +1,77 @@
+              <div class="input-group">
+                <span class="input-group-addon"><i class="fa fa-usd"></i></span>
+                <%= f.hidden_field :user_id, value: current_user.id %>
+                <%= f.text_field :amount_bet, class: 'form-control' %>
```
I would like to see a *hint* in here to show how my entry should be formatted. I tried to bet "one hundred dollars" and I'll I got was a flash telling me to enter a valid amount. If you had a placeholder in this text field to show them the format you are expected I think the user would have a better experience.

LeonTenko committed:
Add validation when placing bet

-Bet option has to be selected;
-Amount of money has to be entered;
-Amount entered should be a number;
-Amount entered can't be empty;

LeonTenko committed:
Add formatting for amount bet

-Now single digit values are padded with another zero.

LeonTenko committed:
Add money formatter for amount bet

-Entered money amount will automatically truncate to two decimal digits.

jacquesporveau commented on app/views/bet_options/_bet_option_radio.html.erb:
```diff
@@ -0,0 +1,9 @@
+  </td>
+  <td class="col col-lg-6"><%= bet_option.option_text %></td>
+  <td class="col col-lg-4">$<%= format('%.2f', bet_option.option_total) %></td>
+  </td>
```
What is this closing `td` tag corresponding to?

jacquesporveau commented on app/views/shared/_bet_display.html.erb:
```diff
@@ -0,0 +1,20 @@
+        </div>
+
+        <div class="class-block">
+          <p><h5 class="card-text"><%= bet.bet %></h5></p>
```
Misusing a header tag. Should not be inside the p.

YarekTenko replied:
Thanks! Hopefully fixed it.

qr8r committed:
Merge pull request #64 from StemboltHQ/users_can_add_bet_options

Users can create BetOptions for Bet
